tests/testsuite/build_script.rs

```diff
@@ -1366,8 +1366,79 @@ fn non_links_cannot_pass_env_vars() {
             r#"
                 use std::env;
                 fn main() {
+                    assert_eq!(env::var("DEP_FOO_FOO").unwrap(), "bar");
+                    assert_eq!(env::var("DEP_FOO_BAR").unwrap(), "baz");
+                    assert_eq!(env::var("CARGO_DEP_A_FOO").unwrap(), "bar");
+                    assert_eq!(env::var("CARGO_DEP_A_BAR").unwrap(), "baz");
+                }
+            "#,
+        )
+        .file(
+            "a/Cargo.toml",
+            r#"
+                [package]
+                name = "a"
+                version = "0.5.0"
+                edition = "2015"
+                authors = []
+                links = "foo"
+                build = "build.rs"
+            "#,
+        )
+        .file("a/src/lib.rs", "")
+        .file(
+            "a/build.rs",
+            r#"
+                use std::env;
+                fn main() {
+                    let lib = env::var("CARGO_MANIFEST_LINKS").unwrap();
+                    assert_eq!(lib, "foo");
+
+                    println!("cargo::metadata=foo=bar");
+                    println!("cargo::metadata=bar=baz");
+                }
+            "#,
+        )
+        .build();
+
+    p.cargo("build -v -Zany-build-script-metadata")
+        .masquerade_as_nightly_cargo(&["any-build-script-metadata"])
+        .run();
+}
+
+#[cargo_test]
+fn non_links_can_pass_env_vars() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.5.0"
+                edition = "2015"
+                authors = []
+                build = "build.rs"
+
+                [dependencies.a]
+                path = "a"
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .file(
+            "build.rs",
+            r#"
+                use std::env;
+                fn main() {
+                    // DEP_<links>_<key> is only allowed for `links` crates
                     assert!(env::var("DEP_FOO_FOO").is_err());
                     assert!(env::var("DEP_FOO_BAR").is_err());
+                    // Make sure DEP_<name-in-toml>_<key> is not present.
+                    // This is not a valid scenario but verify these are not present just incase.
+                    assert!(env::var("DEP_A_FOO").is_err());
+                    assert!(env::var("DEP_A_BAR").is_err());
+
+                    assert_eq!(env::var("CARGO_DEP_A_FOO").unwrap(), "bar");
+                    assert_eq!(env::var("CARGO_DEP_A_BAR").unwrap(), "baz");
```

Comment:
For me, one unresolved question is what to name this env variable
- Technically, `name_key` is ambiguous.  `foo_bar_alice_bob` could be `foo_bar` + `alice_bob`, `foo` + `bar_alice_bob`, etc and *technically* packages could conflict
- Whether we should `envify` the package name and key or not.  iirc we do it in some places but not others.  Overall, I would prefer we would never do it so it matches the name people see in `Cargo.toml`

Reply:
Note that I highlighted these for being called out in the stabilization conversations and do not block this PR.  I don't have concrete ideas to say exactly what should be done but want to raise these for a wider conversation.